drivers/i2c/i2c_ll_stm32_v2.c

```diff
@@ -54,56 +54,71 @@ static inline void msg_init(struct device *dev, struct i2c_msg *msg,
 	}
 }
 
-static inline void msg_done(struct device *dev, unsigned int current_msg_flags)
-{
-	const struct i2c_stm32_config *cfg = DEV_CFG(dev);
-	I2C_TypeDef *i2c = cfg->i2c;
-
-	/* Wait for transfer to complete */
-	while (!LL_I2C_IsActiveFlag_TC(i2c) && !LL_I2C_IsActiveFlag_TCR(i2c)) {
-		;
-	}
-	/* Issue stop condition if necessary */
-	if (current_msg_flags & I2C_MSG_STOP) {
-		LL_I2C_GenerateStopCondition(i2c);
-		while (!LL_I2C_IsActiveFlag_STOP(i2c)) {
-			;
-		}
-		LL_I2C_ClearFlag_STOP(i2c);
-		LL_I2C_DisableReloadMode(i2c);
-	}
-}
-
 #ifdef CONFIG_I2C_STM32_INTERRUPT
 void stm32_i2c_event_isr(void *arg)
 {
 	const struct i2c_stm32_config *cfg = DEV_CFG((struct device *)arg);
 	struct i2c_stm32_data *data = DEV_DATA((struct device *)arg);
 	I2C_TypeDef *i2c = cfg->i2c;
 
-	if (data->current.is_write) {
-		if (data->current.len && LL_I2C_IsActiveFlag_TXIS(i2c)) {
+	/* Interrupts for sending/receiving next byte */
+	if (data->current.len) {
+
+		/* Send next byte */
+		if (data->current.is_write && LL_I2C_IsActiveFlag_TXIS(i2c)) {
 			LL_I2C_TransmitData8(i2c, *data->current.buf);
+
+		/* Receive next byte */
+		} else if (!(data->current.is_write)
+				   && LL_I2C_IsActiveFlag_RXNE(i2c)) {
+			*data->current.buf = LL_I2C_ReceiveData8(i2c);
+
 		} else {
-			LL_I2C_DisableIT_TX(i2c);
 			goto error;
 		}
+
+		data->current.buf++;
+		data->current.len--;
+		if (!data->current.len) {
+			LL_I2C_EnableIT_TC(i2c);
+		}
+
+	/* Interrupts to process end of message */
 	} else {
-		if (data->current.len && LL_I2C_IsActiveFlag_RXNE(i2c)) {
-			*data->current.buf = LL_I2C_ReceiveData8(i2c);
+
+		/* Interrupt after last byte of message
+		 * was transferred
+		 */
+		if (LL_I2C_IsActiveFlag_TC(i2c)
+			|| LL_I2C_IsActiveFlag_TCR(i2c)) {
```

Comment:
Move || on upper line and align LL_I2C_IsActiveFlag_TCR on upper  LL_I2C_IsActiv...